test/test-fs.c

```diff
@@ -3037,6 +3037,54 @@ TEST_IMPL(fs_write_alotof_bufs_with_offset) {
   return 0;
 }
 
+TEST_IMPL(fs_read_dir) {
+  int r;
+  char buf[2];
+  loop = uv_default_loop();
+
+  /* Setup */
+  rmdir("test_dir");
+  r = uv_fs_mkdir(loop, &mkdir_req, "test_dir", 0755, mkdir_cb);
+  ASSERT(r == 0);
+  uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(mkdir_cb_count == 1);
+  /* Setup Done Here */
+
+  /* Get a file descriptor for the directory */
+  r = uv_fs_open(loop,
+                 &open_req1,
+                 "test_dir",
+                 O_RDONLY | O_DIRECTORY,
+                 S_IWUSR | S_IRUSR,
+                 NULL);
+  ASSERT(r > 0);
+  uv_fs_req_cleanup(&open_req1);
+
+  /* Try to read data from the directory */
+  iov = uv_buf_init(buf, sizeof(buf));
+  r = uv_fs_read(NULL, &read_req, open_req1.result, &iov, 1, 0, NULL);
+#if defined(__FreeBSD__) || \
+    defined(__OpenBSD__) || \
+    defined(__NetBSD__)  || \
+    defined(_BSD_SOURCE) || \
+    defined(_AIX)
+  /* As of now, only BSDs and AIX support reading from a directory. */
+  ASSERT(r == 0)
```

Comment:
I would change this to `ASSERT((r >= 0) || r == UV_EISDIR)` as the result depends on the filesystem where the directory lives. I would also add `smartos` here too (see: http://illumos.org/man/read.2) and also `DragonFly`.

Reply:
I included `__sun` and `__DragonFly__`, also made the condition the assertion a bit lenient as you suggested.